src/shared/contexts/buckets.tsx

```diff
@@ -0,0 +1,314 @@
+          `${a.name}`.toLowerCase().localeCompare(`${b.name}`.toLowerCase())
+        )
+        updateCache({
+          loading: RemoteDataState.Done,
+          lastFetch: Date.now(),
+          buckets: [...bucks.user, ...bucks.system, ...bucks.sample],
+        })
+      })
+      .catch(() => {
+        controller.current = null
+        updateCache({
+          loading: RemoteDataState.Error,
+        })
+      })
+  }
+
+  // make sure to fetch buckets on mount
+  useEffect(() => {
+    if (Date.now() - lastFetch > 12 * 60 * 60 * 1000) {
+      fetchBuckets()
+    } else if (loading === RemoteDataState.NotStarted) {
+      fetchBuckets()
+    }
+  }, [])
+
+  const createBucket = (bucket: ExtendedBucket) => {
+    bucket.orgID = scope.org
+    let schemas = []
+    if (bucket.schemas && bucket.schemas.length) {
+      bucket.schemaType = 'explicit'
+      schemas = bucket.schemas
+    } else {
+      bucket.schemaType = 'implicit'
+    }
+
+    delete bucket.schemas
+
+    const headers = {
+      'Content-Type': 'application/json',
+      'Accept-Encoding': 'gzip',
+    }
+
+    if (scope?.token) {
+      headers['Authorization'] = `Token ${scope.token}`
+    }
+
+    fetch(`${scope?.region}/api/v2/buckets`, {
+      method: 'POST',
+      headers,
+      body: JSON.stringify(bucket),
+    })
+      .then(response => {
+        return response.json()
+      })
+      .then(response => {
+        addBucket(response)
+
+        return Promise.all(
+          schemas.map(schema =>
+            fetch(
+              `${scope?.region}/api/v2/buckets/${response.id}/schema/measurements`,
+              {
+                method: 'POST',
+                headers,
+                body: JSON.stringify(schema),
+              }
+            ).then(resp => {
+              if (resp.status !== 201) {
+                dispatch(
+                  notify(
+                    measurementSchemaAdditionFailed(response.name, schema.name)
+                  )
+                )
+              } else {
+                dispatch(
+                  notify(
+                    measurementSchemaAdditionSuccessful(
+                      response.name,
+                      schema.name
+                    )
+                  )
+                )
+              }
+            })
+          )
+        )
+      })
+  }
+
+  const addBucket = (bucket: Bucket) => {
+    const bucks = buckets.reduce(
+      (acc, curr) => {
+        if (!acc[curr.type]) {
+          acc[curr.type] = []
+        }
+        acc[curr.type].push(curr)
+        return acc
+      },
+      {[bucket.type]: [bucket]}
+    )
+
+    bucks.system.sort((a, b) =>
+      `${a.name}`.toLowerCase().localeCompare(`${b.name}`.toLowerCase())
+    )
+    bucks.user.sort((a, b) =>
+      `${a.name}`.toLowerCase().localeCompare(`${b.name}`.toLowerCase())
+    )
+    bucks.sample.sort((a, b) =>
+      `${a.name}`.toLowerCase().localeCompare(`${b.name}`.toLowerCase())
+    )
```

Comment:
there is only one bucket passed into `addBucket` function, so line 282 will construct an object with one key instead of three keys (i.e `system`, `user`, `sample`). it seems like any two of these lines will have a run-time error like `bucks.system` is undefined

Reply:
274 is operating on the bucket cache (`buckets`) and adding the new bucket (`bucket`) by prepopulating it into the result set of that reduce function with line 282.

Reply:
turns out that notebooks is calling this with a cold cache before the bucket list has loaded. only affects existing notebooks before a new notebook has been created